core/consensus/babe/primitives/src/digest.rs

```diff
@@ -93,6 +88,30 @@ impl BabePreDigest {
 	}
 }
 
+/// Get the slot.
+pub fn get_slot<H: Header>(header: &H) -> Result<SlotNumber, &str>
+	where DigestItemForHeader<H>: CompatibleDigestItem,
+{
+	find_pre_digest::<H, RawBabePreDigest>(header)
+		.map(|raw_pre_digest| raw_pre_digest.slot_number())
+}
+
+/// Extract the BABE pre digest from the given header. Pre-runtime digests are
+/// mandatory, the function will return `Err` if none is found.
+pub fn find_pre_digest<H: Header, D: Codec>(header: &H) -> Result<D, &str>
```

Comment:
it's strange to return a `str` with the lifetime of the given header as opposed to a `&'static str`.